postgis/lwgeom_box3d.c

```diff
@@ -206,62 +206,194 @@ Datum BOX3D_to_LWGEOM(PG_FUNCTION_ARGS)
 		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmax;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymax;
+		pt.z = box->zmax;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 		pt.x = box->xmin;
 		pt.y = box->ymax;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+
+		ppa[0] = pa;
+		lwpoly = lwpoly_construct(SRID_UNKNOWN, NULL, 1, ppa);
+		result = geometry_serialize(lwpoly_as_lwgeom(lwpoly));
+		lwpoly_free(lwpoly);
+	}
+	else if (box->ymin == box->ymax)
+	{
+		LWPOLY *lwpoly;
+		POINTARRAY **ppa = lwalloc(sizeof(POINTARRAY*));
+
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmax;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmax;
+		pt.y = box->ymin;
+		pt.z = box->zmax;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmax;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+
+		ppa[0] = pa;
+		lwpoly = lwpoly_construct(SRID_UNKNOWN, NULL, 1, ppa);
+		result = geometry_serialize(lwpoly_as_lwgeom(lwpoly));
+		lwpoly_free(lwpoly);
+	}
+	else if (box->zmin == box->zmax)
+	{
+		LWPOLY *lwpoly;
+		POINTARRAY **ppa = lwalloc(sizeof(POINTARRAY *));
+
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymax;
+		pt.z = box->zmin;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 		pt.x = box->xmax;
 		pt.y = box->ymax;
+		pt.z = box->zmin;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 		pt.x = box->xmax;
 		pt.y = box->ymin;
+		pt.z = box->zmin;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 		pt.x = box->xmin;
 		pt.y = box->ymin;
+		pt.z = box->zmin;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 
+		ppa[0] = pa;
+		lwpoly = lwpoly_construct(SRID_UNKNOWN, NULL, 1, ppa);
 		result = geometry_serialize(lwpoly_as_lwgeom(lwpoly));
-		
+		lwpoly_free(lwpoly);
+	}
+	else
+	{
+		POINT4D points[8];
+		LWGEOM **geoms = NULL;
+		LWGEOM *geom = NULL;
+		int ngeoms = 6;
+
+		/* Initialize the 8 vertices of the box */
+		points[0] = (POINT4D) { box->xmin, box->ymin, box->zmin };
+		points[1] = (POINT4D) { box->xmin, box->ymax, box->zmin };
+		points[2] = (POINT4D) { box->xmax, box->ymax, box->zmin };
+		points[3] = (POINT4D) { box->xmax, box->ymin, box->zmin };
+		points[4] = (POINT4D) { box->xmin, box->ymin, box->zmax };
+		points[5] = (POINT4D) { box->xmin, box->ymax, box->zmax };
+		points[6] = (POINT4D) { box->xmax, box->ymax, box->zmax };
+		points[7] = (POINT4D) { box->xmax, box->ymin, box->zmax };
+
+		geoms = (LWGEOM **) lwalloc(sizeof(LWGEOM *) * ngeoms);
```

Comment:
The points var is already stack memory

geoms can be made stack easily too (switch to LWGEOM *geoms[6], and some fixes to the destructor).

The lwcollection (geom here) can be entirely build in stack memory, but I think it's too much knowledge of the datatype.

I don't really see how to avoid heap allocations in lwpoly_construct_rectangle without redefining a pointarray and LWPOLY from scratch like the lwcollection. I think it's the most expensive part of this code path though.

A quick example with geoms and the lwcollection in stack:
```C
		POINT4D points[8];
		LWGEOM *geoms[6];
		LWCOLLECTION collection;
		int ngeoms = 6, i;

		/* Initialize the 8 vertices of the box */
		points[0] = (POINT4D) { box->xmin, box->ymin, box->zmin };
		points[1] = (POINT4D) { box->xmin, box->ymax, box->zmin };
		points[2] = (POINT4D) { box->xmax, box->ymax, box->zmin };
		points[3] = (POINT4D) { box->xmax, box->ymin, box->zmin };
		points[4] = (POINT4D) { box->xmin, box->ymin, box->zmax };
		points[5] = (POINT4D) { box->xmin, box->ymax, box->zmax };
		points[6] = (POINT4D) { box->xmax, box->ymax, box->zmax };
		points[7] = (POINT4D) { box->xmax, box->ymin, box->zmax };

		/* add bottom polygon */
		geoms[0] = (lwpoly_as_lwgeom(lwpoly_construct_rectangle(&points[0],
						&points[1], &points[2], &points[3])));
		/* add top polygon */
		geoms[1] = (lwpoly_as_lwgeom(lwpoly_construct_rectangle(&points[4],
						&points[5], &points[6], &points[7])));
		/* add left polygon */
		geoms[2] = (lwpoly_as_lwgeom(lwpoly_construct_rectangle(&points[0],
						&points[4], &points[5], &points[1])));
		/* add right polygon */
		geoms[3] = (lwpoly_as_lwgeom(lwpoly_construct_rectangle(&points[3],
						&points[7], &points[6], &points[2])));
		/* add back polygon */
		geoms[4] = (lwpoly_as_lwgeom(lwpoly_construct_rectangle(&points[0],
						&points[4], &points[7], &points[3])));
		/* add front polygon */
		geoms[5] = (lwpoly_as_lwgeom(lwpoly_construct_rectangle(&points[1],
						&points[5], &points[6], &points[2])));

		collection.type = POLYHEDRALSURFACETYPE;
		collection.flags = gflags(1, 0, 0);
		FLAGS_SET_BBOX(collection.flags, 0);
		collection.bbox = NULL;
		collection.srid = SRID_UNKNOWN;
		collection.ngeoms = ngeoms;
		collection.maxgeoms = ngeoms;
		collection.geoms = geoms;
		FLAGS_SET_SOLID(collection.flags, 1);

		result = geometry_serialize((LWGEOM *) &collection);

		for (i=0; i<ngeoms; i++)
			lwpoly_free((LWPOLY *) geoms[i]);
```

Was is something like that you had in mind?